### PR TITLE
Add kubeval blueprint

### PR DESCRIPTION
**Pull request by prachirp**

### CODE REVIEW

**Comment by prachirp** [comment]
cc @frankfarzan PTAL

File: examples/kubeval/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Kubeval
+
+The `kubeval` KRM config function validates configuration using kubeval. This
```
Comment (frankfarzan) [inline examples/kubeval/README.md:3]: What's kubeval? Why does the user need to care? Provide context, don't assume the reader is familiar with all these proper nouns in the catalog.

File: examples/kubeval/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+example invokes the kubeval function using declarative configuration.
+
+## Function invocation
+
+Get this example and try it out by running the following commands:
+
+```sh
+kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval .
+kpt fn run kubeval --network --results-dir /tmp
```
Comment (frankfarzan) [inline examples/kubeval/README.md:12]: why does this need to have `--results-dir`? KISS principle: just print to stdout.

File: examples/kubeval/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+cat /tmp/results-0.yaml
+```
+
+They contain the following error:
+
+```sh
+Invalid type. Expected: [integer,null], given: string
+```
+
+The error comes from the `bob` resource in `configs/example-config.yaml`.
```
Comment (frankfarzan) [inline examples/kubeval/README.md:30]: let's standardize having a comment detailing the fix

File: examples/kubeval/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+The `kubeval` KRM config function validates configuration using kubeval. This
+example invokes the kubeval function using declarative configuration.
+
+## Function invocation
+
+Get this example and try it out by running the following commands:
+
+```sh
+kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval .
```
Comment (frankfarzan) [inline examples/kubeval/README.md:11]: we discussed this should also show how to hit a live cluster

File: examples/kubeval/README.md (hunk 1 above)
Comment (prachirp) [reply examples/kubeval/README.md]: Updated the description

File: examples/kubeval/README.md (hunk 2 above)
Comment (prachirp) [reply examples/kubeval/README.md]: Removed

File: examples/kubeval/README.md (hunk 3 above)
Comment (prachirp) [reply examples/kubeval/README.md]: Done

File: examples/kubeval/README.md (hunk 4 above)
Comment (prachirp) [reply examples/kubeval/README.md:12]: Kubeval doesn't hit live clusters. It accepts args for --kubernetes-version and for --schema-location and uses a k8s version's schema or the schema passed in to validate resources against. I use --kubernetes-version 1.18.0 in this blueprint.

File: examples/kubeval/fn-config.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  name: my-func-config
+  annotations:
+    config.k8s.io/function: |
+      container:
+        image: gcr.io/kpt-functions/kubeval
+        network:
+          required: true
+    config.kubernetes.io/local-config: 'true'
+data:
+  schema_location: https://kubernetesjsonschema.dev/
```
Comment (frankfarzan) [inline examples/kubeval/fn-config.yaml:13]: newline

File: examples/kubeval/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+kpt fn run kubeval --network
+```
+
+## Expected Results
+
+The function outputs the following error:
+
+```sh
+[ERROR] Invalid type. Expected: [integer,null], given: string in object 'v1/ReplicationController//bob' in file configs/example-config.yaml
```
Comment (frankfarzan) [inline examples/kubeval/README.md:21]: Why isn't the path to the problematic field part of the error message?

This example has `spec.replicas` in the output message:

https://www.kubeval.com/#json

File: examples/kubeval/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# Kubeval
+
+The `kubeval` KRM config function validates Kubernetes resources against their
```
Comment (frankfarzan) [inline examples/kubeval/README.md:3]: nit: Remove unnecessary filler:

`Validates Kubernetes resources against their Kubernetes OpenAPI definition using [kubeval]`

Link to kubeval docs.

File: examples/kubeval/README.md (hunk 4 above)
Comment (frankfarzan) [reply examples/kubeval/README.md:12]: Nice

File: examples/kubeval/README.md (hunk 6 above)
Comment (prachirp) [reply examples/kubeval/README.md:21]: This is related to https://github.com/GoogleContainerTools/kpt/issues/719. When kpt prints results to console, it doesn't print the full results. I will have this blueprint use `--results-dir` and the results file clearly shows the error comes from field path `spec.replicas`.

File: examples/kubeval/README.md (hunk 7 above)
Comment (prachirp) [reply examples/kubeval/README.md]: Done

File: examples/kubeval/fn-config.yaml (hunk 5 above)
Comment (prachirp) [reply examples/kubeval/fn-config.yaml]: Done

File: examples/kubeval/README.md (hunk 6 above)
Comment (frankfarzan) [reply examples/kubeval/README.md]: Why does omitting tags impact this? `spec.replicas` is part of field path not a tag.

Comment (prachirp) [reply examples/kubeval/README.md]: I updated the issue. Kpt only includes the following fields to output to the console: message, severity, resourceRef, and file path. Basically everything else is omitted. As a result we need to add `--results-dir` and view the yaml for the full results.

Comment (prachirp) [reply examples/kubeval/README.md]: Included the full results which show `spec.replicas`

File: examples/kubeval/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+to. Check the results:
+
+```sh
+cat /tmp/results-0.yaml
+```
+
+They contain the following error message:
+
+```sh
+Invalid type. Expected: [integer,null], given: string
```
[the remark this answers is not in the document]
Comment (prachirp) [reply examples/kubeval/README.md]: Included the full results which show `spec.replicas`

File: examples/kubeval/README.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+This example invokes the kubeval function against Kubernetes v1.18.0 using
+declarative configuration.
+
+## Function invocation
+
+Get this example and try it out by running the following commands:
+
+```sh
+kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval .
+kpt fn run kubeval --network --results-dir /tmp
```
Comment (frankfarzan) [inline examples/kubeval/README.md:15]: Is there a e2e test for `--results-dir`?

**Comment by frankfarzan** [comment]
Add a TODO to remove --results-dir from this example.

File: examples/kubeval/README.md (hunk 9 above)
Comment (prachirp) [reply examples/kubeval/README.md:17]: Added e2e tests for --results-dir and added TODOs linking to issues about removing --results-dir from examples.

**Comment by prachirp** [comment]
cc @frankfarzan PTAL

File: tests/istioctl_analyze.sh [hunk 10]
```diff
@@ -86,5 +86,11 @@ assert_contains_string error.txt "Referenced gateway not found"
 
 testcase "kpt_istioctl_analyze_declarative_example"
 kpt pkg get "$CATALOG_REPO"/examples/istioctl-analyze .
+kpt fn run istioctl-analyze 2>err.txt || true
+assert_contains_string err.txt "Schema validation error: gateway must have at least one server"
+
+# TODO: Remove after fixing https://github.com/GoogleContainerTools/kpt/issues/983
```
Comment (frankfarzan) [inline tests/istioctl_analyze.sh:92]: The assertion is same for both tests (unlike kubeval). Revert changes to istioctl_analyze.sh, you don't want to be testing kpt CLI here, just the functions.

Comment (prachirp) [reply tests/istioctl_analyze.sh]: Done

**Review by frankfarzan — APPROVED** [review]

**Pull request merged**